File: src/models/game.ts

// 遊戲狀態枚舉
import {MessageType} from "../components/game/GameMessages";

export enum GameStatus {
    WAITING = 'WAITING',
    PLAYING = 'PLAYING',
    FINISHED = 'FINISHED'
}

// 行動類型枚舉
export enum ActionType {
    SECRET = 'SECRET',
    DISCARD = 'DISCARD',
    GIFT = 'GIFT',
    COMPETE = 'COMPETE'
}

// 藝妓青睞狀態
export enum FavorStatus {
    NEUTRAL = 'NEUTRAL',
    PLAYER1 = 'PLAYER1',
    PLAYER2 = 'PLAYER2'
}

// 卡牌接口
export interface Card {
    id: string;
    geishaId: string;
    geishaName: string;
    status: 'IN_DECK' | 'IN_HAND' | 'ALLOCATED' | 'SECRET' | 'DISCARDED';
    ownerId?: string;
}

// 藝妓接口
export interface Geisha {
    id: string;
    name: string;
    charm: number;
    favor: FavorStatus;
}

// 玩家接口
export interface Player {
    id: string;
    name: string;
    isMyTurn: boolean;
    hand: Card[];
    usedActions: ActionType[];
    allocatedCards: {[geishaId: string]: Card[]};
    score: number;
}

// 遊戲接口
export interface Game {
    id: string;
    status: GameStatus;
    currentPlayerId: string;
    players: {[playerId: string]: Player};
    geishas: {[geishaId: string]: Geisha};
    round: number;
    messages: GameMessage[];
}

// 遊戲訊息
export interface GameMessage {
    id: string;
    type: MessageType;
    text: string;
    timestamp: number;
    playerId?: string;
    playerName?: string;
    actionType?: string;
    details?: any;
}

// 行動請求
export interface ActionRequest {
    gameId: string;
    actionType: ActionType;
    cardIds: string[];
    groupings?: string[][];  // 用於獻禮和競爭行動
}

File: hanamikoji-backend/app/schemas/game.py
"""遊戲相關的Pydantic模型"""

from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any
from enum import Enum


class GameStatus(str, Enum):
    """遊戲狀態"""
    WAITING = "WAITING"
    PLAYING = "PLAYING" 
    FINISHED = "FINISHED"


class ActionType(str, Enum):
    """行動類型"""
    SECRET = "SECRET"      # 秘密保留
    DISCARD = "DISCARD"    # 棄牌
    GIFT = "GIFT"          # 獻禮
    COMPETE = "COMPETE"    # 競爭


class FavorStatus(str, Enum):
    """青睞狀態"""
    NEUTRAL = "NEUTRAL"
    PLAYER1 = "PLAYER1"
    PLAYER2 = "PLAYER2"


class GiftCard(BaseModel):
    """禮物卡模型"""
    id: str
    geisha_id: str
    item_name: str
    charm_value: int
    status: str = "IN_HAND"
    owner_id: Optional[str] = None


class Geisha(BaseModel):
    """藝妓模型"""
    id: str
    name: str
    charm: int
    gift_item: str
    description: Optional[str] = None
    favor: FavorStatus = FavorStatus.NEUTRAL
    allocated_gifts: Dict[str, List[GiftCard]] = Field(default_factory=dict)


class Player(BaseModel):
    """玩家模型"""
    id: str
    name: str
    hand_cards: List[GiftCard] = Field(default_factory=list)
    used_actions: List[ActionType] = Field(default_factory=list)
    secret_cards: List[GiftCard] = Field(default_factory=list)
    allocated_gifts: Dict[str, List[GiftCard]] = Field(default_factory=dict)
    score: int = 0
    is_current_player: bool = False


class GameCreateRequest(BaseModel):
    """創建遊戲請求"""
    player1_name: str = Field(..., min_length=1, max_length=50)
    player2_name: str = Field(..., min_length=1, max_length=50)


class ActionRequest(BaseModel):
    """遊戲動作請求"""
    player_id: str
    action_type: ActionType
    card_ids: List[str]
    target_geisha_id: Optional[str] = None
    groupings: Optional[List[List[str]]] = None


class GameMessage(BaseModel):
    """遊戲訊息"""
    id: str
    type: str
    text: str
    timestamp: int
    player_id: Optional[str] = None
    player_name: Optional[str] = None
    action_type: Optional[str] = None
    details: Optional[Dict[str, Any]] = None


class GameStateResponse(BaseModel):
    """遊戲狀態回應"""
    game_id: str
    status: GameStatus
    current_player_id: str
    round_number: int
    players: Dict[str, Player]
    geishas: List[Geisha]
    messages: List[GameMessage] = Field(default_factory=list)
    winner: Optional[str] = None


class GameStatusResponse(BaseModel):
    """遊戲狀態簡要回應"""
    game_id: str
    status: GameStatus
    current_player_id: str
    round_number: int
    player_names: List[str]
    created_at: str
    winner: Optional[str] = None


class GameListItem(BaseModel):
    """遊戲列表項目"""
    game_id: str
    status: GameStatus
    player_names: List[str]
    created_at: str
    current_round: int